examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/pkg/websocket/client.go

```diff
@@ -0,0 +1,65 @@
+package websocket
+
+import (
+	"crypto/tls"
+	"fmt"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend/log"
+	"golang.org/x/net/websocket"
+)
+
+var Logger = log.DefaultLogger
+
+type Client struct {
+	conn *websocket.Conn
+}
+
+type Options struct {
+	URI string `json:"uri"`
+}
+
+func NewClient(serverURL string) (*Client, error) {
+	config, err := websocket.NewConfig(serverURL, "http://localhost")
+	if err != nil {
+		return nil, fmt.Errorf("error creating WebSocket config: %v", err)
+	}
+	// use self-signed certificate for this test example
+	config.TlsConfig = &tls.Config{InsecureSkipVerify: true}
+	conn, err := websocket.DialConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to WebSocket server: %v", err)
+	}
+
+	return &Client{conn: conn}, nil
+}
+
+func (c *Client) Read() <-chan string {
+	messages := make(chan string)
+	go func() {
+		defer close(messages)
+		for {
+			var message string
+			err := websocket.Message.Receive(c.conn, &message)
+			if err != nil {
+				Logger.Info("Error reading message: %v", err)
+				continue
+			}
+			messages <- message
+		}
+	}()
+	return messages
+}
```

Comment:
Nice catch! I solve this closing the connection from the caller when `ctx.Done()`. This will case an error on `websocket.Message.Receive(c.conn, &message)`, because the connection is close, then the loop will continue to the next case which is the `ctx.Done()`.
Here I changed a little bit your suggestion and `return` instead of `break` to avoid the `outer` label

Reply:
After some changes in the code, I've introduced one connection per query (like it was done in mqtt plugin) to fulfill Andres concerns about changing the datasource. 
Doing that I was having situation where the connection was close, to I would receive an err in `websocket.Message.Receive` but the `ctx.Done()` fired, so I've introduced a new channel `stopChan` that would be closed when the connection is closed and that would stop the loop